server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java

```diff
@@ -0,0 +1,174 @@
+            boolean hasDocs = false;
+            for (var loader : loaders) {
+                hasDocs |= loader.advanceToDoc(docId);
+            }
+
+            this.hasValue |= hasDocs;
+            return hasDocs;
+        };
+    }
+
+    @Override
+    public boolean hasValue() {
+        return hasValue;
+    }
+
+    @Override
+    public void write(XContentBuilder b) throws IOException {
+        var totalCount = Arrays.stream(parts).mapToLong(SyntheticFieldLoaderLayer::valueCount).sum();
+
+        if (totalCount == 0) {
+            return;
+        }
+
+        if (totalCount == 1) {
+            b.field(fieldName);
+            for (var part : parts) {
+                part.write(b);
+            }
+            return;
+        }
+
+        b.startArray(fieldName);
+        for (var part : parts) {
+            part.write(b);
+        }
+        b.endArray();
+    }
+
+    @Override
+    public String fieldName() {
+        return this.fullFieldName;
+    }
+
+    /**
+     * Represents one layer of loading synthetic source values for a field
+     * as a part of {@link CompositeSyntheticFieldLoader}.
+     * <br>
+     * Note that the contract of {@link SourceLoader.SyntheticFieldLoader#write(XContentBuilder)}
+     * is slightly different here since it only needs to write field values without encompassing object or array.
+     */
+    public interface SyntheticFieldLoaderLayer extends SourceLoader.SyntheticFieldLoader {
+        /**
+         * Number of values that this loader will write.
+         * @return
+         */
+        long valueCount();
+    }
+
+    /**
+     * Layer that loads malformed values stored in a dedicated field with a conventional name.
+     * @see IgnoreMalformedStoredValues
+     */
+    public static class MalformedValuesLayer implements SyntheticFieldLoaderLayer {
+        private final String fieldName;
+        private List<Object> values;
+
+        public MalformedValuesLayer(String fieldName) {
+            this.fieldName = fieldName + "._ignore_malformed";
+            this.values = emptyList();
+        }
+
+        @Override
+        public long valueCount() {
+            return values.size();
+        }
+
+        @Override
+        public Stream<Map.Entry<String, StoredFieldLoader>> storedFieldLoaders() {
+            return Stream.of(Map.entry(fieldName, values -> this.values = values));
+        }
+
+        @Override
+        public DocValuesLoader docValuesLoader(LeafReader leafReader, int[] docIdsInLeaf) throws IOException {
+            return null;
+        }
+
+        @Override
+        public boolean hasValue() {
+            return values.isEmpty() == false;
+        }
+
+        @Override
+        public void write(XContentBuilder b) throws IOException {
+            for (Object v : values) {
+                if (v instanceof BytesRef r) {
+                    XContentDataHelper.decodeAndWrite(b, r);
+                } else {
+                    b.value(v);
```

Comment:
What's the use case for this one? I thought malformed values are always encoded.

Reply:
If it's f.e. text we skip encoding in some fields. This is for compatibility with existing code.